spring-webflux/src/main/java/org/springframework/web/reactive/socket/adapter/JettyWebSocketSession.java

```diff
@@ -51,52 +70,100 @@ public JettyWebSocketSession(Session session, HandshakeInfo info, DataBufferFact
+		this.sink.next(message);
 
-	@Override
-	protected boolean canSuspendReceiving() {
-		// Jetty 12 TODO: research suspend functionality in Jetty 12
-		return false;
+		boolean demand = false;
+		this.lock.lock();
+		try {
+			if (!this.awaitingDemand) {
+				throw new IllegalStateException();
+			}
+			this.awaitingDemand = false;
+			if (this.requested > 0) {
+				this.requested--;
+				this.awaitingDemand = true;
+				demand = true;
+			}
+		}
+		finally {
+			this.lock.unlock();
+		}
+
+		if (demand) {
+			getDelegate().demand();
+		}
 	}
 
-	@Override
-	protected void suspendReceiving() {
+	void handleError(Throwable ex) {
 	}
 
-	@Override
-	protected void resumeReceiving() {
+	void handleClose(CloseStatus closeStatus) {
+		this.closeStatusSink.tryEmitValue(closeStatus);
+		this.sink.complete();
 	}
 
-	@Override
-	protected boolean sendMessage(WebSocketMessage message) throws IOException {
-		DataBuffer dataBuffer = message.getPayload();
-		Session session = getDelegate();
-		if (WebSocketMessage.Type.TEXT.equals(message.getType())) {
-			getSendProcessor().setReadyToSend(false);
-			String text = dataBuffer.toString(StandardCharsets.UTF_8);
-			session.sendText(text, new SendProcessorCallback());
-		}
-		else {
-			if (WebSocketMessage.Type.BINARY.equals(message.getType())) {
-				getSendProcessor().setReadyToSend(false);
+	void onHandlerError(Throwable error) {
+		getDelegate().close(StatusCode.SERVER_ERROR, error.getMessage(), new Callback() {
+			@Override
+			public void succeed() {
+				if (JettyWebSocketSession.this.handlerCompletionSink != null) {
+					JettyWebSocketSession.this.handlerCompletionSink.tryEmitError(error);
+				}
 			}
-			try (DataBuffer.ByteBufferIterator iterator = dataBuffer.readableByteBuffers()) {
-				while (iterator.hasNext()) {
-					ByteBuffer byteBuffer = iterator.next();
-					switch (message.getType()) {
-						case BINARY -> session.sendBinary(byteBuffer, new SendProcessorCallback());
-						case PING -> session.sendPing(byteBuffer, new SendProcessorCallback());
-						case PONG -> session.sendPong(byteBuffer, new SendProcessorCallback());
-						default -> throw new IllegalArgumentException("Unexpected message type: " + message.getType());
-					}
+
+			@Override
+			public void fail(Throwable ex) {
+				if (JettyWebSocketSession.this.handlerCompletionSink != null) {
+					error.addSuppressed(ex);
+					JettyWebSocketSession.this.handlerCompletionSink.tryEmitError(error);
 				}
 			}
-		}
-		return true;
+		});
+	}
+
+	void onHandleComplete() {
+		getDelegate().close(StatusCode.NORMAL, null, new Callback() {
+			@Override
+			public void succeed() {
+				if (JettyWebSocketSession.this.handlerCompletionSink != null) {
+					JettyWebSocketSession.this.handlerCompletionSink.tryEmitEmpty();
+				}
+			}
+
+			@Override
+			public void fail(Throwable ex) {
+				if (JettyWebSocketSession.this.handlerCompletionSink != null) {
+					JettyWebSocketSession.this.handlerCompletionSink.tryEmitEmpty();
```

Comment:
The error is not relayed here?

Reply:
I don't think `handlerCompletionSink` cares about the failed result. I think I even got a failing test when I tried emitting the error to the `handlerCompletionSink`.

In fact the `AbstractListenerWebSocketSession` does this
```java
if (this.handlerCompletionSink != null) {
	// Ignore result: can't overflow, ok if not first or no one listens
	this.handlerCompletionSink.tryEmitEmpty();
}
close();
```

Where close returns a Mono<Void> which is completely ignored.

Reply:
Maybe do the `tryEmitEmpty()` before calling `getDelegate().close()`?

Reply:
done